### PR TITLE
Synchronize RRef.to_here() CUDA Streams properly 2

### PR DESCRIPTION
**Pull request by pbelevich**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55436 Synchronize RRef.to_here() CUDA Streams properly 2**

### CODE REVIEW

File: torch/csrc/distributed/rpc/rref_impl.cpp [hunk 1]
```diff
@@ -2,10 +2,12 @@
 #include <fmt/format.h>
 #include <torch/csrc/distributed/autograd/rpc_messages/rpc_with_autograd.h>
 #include <torch/csrc/distributed/autograd/utils.h>
+#include <torch/csrc/distributed/rpc/macros.h>
 #include <torch/csrc/distributed/rpc/profiler/remote_profiler_manager.h>
 #include <torch/csrc/distributed/rpc/rref_context.h>
 #include <torch/csrc/distributed/rpc/rref_impl.h>
 #include <torch/csrc/distributed/rpc/rref_proto.h>
+#include <torch/csrc/distributed/rpc/tensorpipe_utils.h>
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/rref_impl.cpp:10]: Just a quick comment. Please feel free to ignore if this is already in your plan or if I was wrong. 

`tensorpipe_*` files are in `libtorch_python_distributed_sources` and `rref_*` files are in `libtorch_distributed_sources`. In general `libtorch_python` files can depend on `libtorch` files, but not the other way around. 

So, we might need to extract the ctx implementation from `tensorpipe_utils.h` and put that in, maybe, `rpc/utils.h`. And then both TensorPipe and RRef can use that implementation.

see build_variables.bzl for file categorization: https://github.com/pytorch/pytorch/blob/b39eeb07ed2f60a27b60ad32d172b58d31554282/tools/build_variables.bzl

**Review by mrshenli — COMMENTED** [review]
quick comment :)

We need to add some unit tests to `rpc_test.py`. Since this will be TensorPipe only, let's add those test functions to `class TensorPipeAgentRpcTest`.

File: torch/csrc/distributed/rpc/rref_impl.h [hunk 2]
```diff
@@ -395,6 +399,28 @@ class TORCH_API OwnerRRef final : public RRef {
...
+  // otherwise does nothing.
+  void recordAllStreams(std::shared_ptr<LazyStreamContext> ctx);
+
+  // Blocks all streams in the context on all events previously stored in
+  // the current OwnerRRef instance if CUDA is available,
+  // otherwise does nothing.
+  void blockAllStreams(std::shared_ptr<LazyStreamContext> ctx);
+ private:
+
+ #ifdef USE_CUDA_NOT_ROCM
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/rref_impl.h:415]: indent

File: torch/csrc/distributed/rpc/rref_impl.h [hunk 3]
```diff
@@ -395,6 +399,28 @@ class TORCH_API OwnerRRef final : public RRef {
...
+  // a storage for CUDA events for synchronization.
+  std::vector<at::cuda::CUDAEvent> cudaEvents_;
+#else
+  // if CUDA is not available we still need a dummy field
+  // to prevent segfaults because of different class size
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/rref_impl.h:420]: Curious, why the different class sizes could be a problem? I thought these are compile time flags, which means when CUDA is not available, it's same as this class does not have this field, no?

File: torch/csrc/distributed/rpc/rref_impl.h (hunk 2 above)
Comment (pbelevich) [reply torch/csrc/distributed/rpc/rref_impl.h:415]: I always thought that [the directives should start at the beginning of the line](https://google.github.io/styleguide/cppguide.html#Preprocessor_Directives)

**Comment by pbelevich** [comment]
Close in favor on https://github.com/pytorch/pytorch/pull/54932

File: torch/csrc/distributed/rpc/rref_impl.h (hunk 3 above)
Comment (pbelevich) [reply torch/csrc/distributed/rpc/rref_impl.h:420]: The practice shows that this header is included to cpp files with CUDA and without and it produces segfaults

**Pull request closed without merging**